text/0070-pd-record-historical-hot-region.md

```diff
@@ -0,0 +1,202 @@
+
+### Table Header Design
+1. New table header
+  ```SQL
+  > USE information_schema;
+  > DESC tidb_hot_regions_history;
+  +-------------+-------------+------+------+---------+-------+
+  | Field       | Type        | Null | Key  | Default | Extra |
+  +-------------+-------------+------+------+---------+-------+
+  | UPDATE_TIME | varchar(64) | YES  |      | NULL    |       | // new
+  | DB_NAME     | varchar(64) | YES  |      | NULL    |       |
+  | TABLE_NAME  | varchar(64) | YES  |      | NULL    |       |
+  | TABLE_ID    | bigint(21)  | YES  |      | NULL    |       |
+  | INDEX_NAME  | varchar(64) | YES  |      | NULL    |       |
+  | INDEX_ID    | bigint(21)  | YES  |      | NULL    |       |
+  | REGION_ID   | bigint(21)  | YES  |      | NULL    |       | 
+  | STORE_ID    | bigint(21)  | YES  |      | NULL    |       | // new
+  | PEER_ID     | bigint(21)  | YES  |      | NULL    |       | // new
+  | TYPE        | varchar(64) | YES  |      | NULL    |       |
+  | HOT_DEGREE  | bigint(21)  | YES  |      | NULL    |       | // rename to HOT_DEGREE
+  | FLOW_BYTES  | double      | YES  |      | NULL    |       |
+  | KEY_RATE    | double      | YES  |      | NULL    |       | // new
+  | QUERY_RATE  | double      | YES  |      | NULL    |       | // new
+  +-------------+-------------+------+------+---------+-------+
+  | REGION_COUNT| bigint(21)  | YES  |      | NULL    |       | // Deleted fields
+  ```
+  * Add `UPDATE_TIME` to support history.
+  * Add `STORE_ID` and `PEER_ID`to track the machine of region.
+  * Rename `MAX_HOT_DEGREE`  to `HOT_DEGREE` for precise meaning in history scenario.
+  * Add `KEY_RATE` and `QUERY_RATE` for future expansion in hotspot determination dimensions.
+  * Remove `REGION_COUNT` for disuse and repeat with `STORE_ID`.
+
+2. Data size estimation 
+
+    one record size: 8 * 8B(bitint) + 1 * 8B(datetime) + 4 * 64B(varchar(64)) = 328B
+    Below table show data size per day and per month in 5,10,15 minutes record interval respectively given the maximum number of hotspot regions  is 1000:
+
+  | Record Length (B)| Time Interval (Min)  | Data Size Per Day (MB)   | Data Size Per Month (MB)   |
+  | ---------------- | -------------------- | ------------------------ | -------------------------- |
+  | 328              | 5                    | 90.08789063              | 2702.636719                |
+  | 328              | 10                   | 45.04394531              | 1351.318359                |
+  | 328              | 15                   | 30.02929688              | 900.8789063                |
+
+### Design in PD
+1. Timing write：
+
+     The leader of PD will periodically encrypt  `START_KEY` and `END_KEY`  in data,and write hotspot region data into `LevelDB`.The write interval can be configured. The write fieldes are: `REGION_ID`, `TYPE`,  `HOT_DEGREE`, `FLOW_BYTES`, `KEY_RATE`, `QUERY_RATE`, `STORE_ID`, `PEER_ID`, `UPDATE_TIME`, `START_KEY`,  `END_KEY`.
```

Comment:
Does PD currently uses `LevelDB`? Maybe we should not add more dependency, especially these requiring cgo, in pd...

Reply:
Yes,`LevelDB` is used  to store region info, [code](https://github.com/tikv/pd/blob/20b00a7c4933e307c5423da0c5047250259e19e9/server/core/region_storage.go#L60).